cep-recipe-jinja.md

```diff
@@ -0,0 +1,320 @@
+
+```yaml
+${{ compiler('c') }}
+```
+
+This would pull in the c_compiler and c_compiler_version from the variant config. The compiler function suffixes `{lang}_compiler` with the `target_platform` to render to something such as:
+
+```
+gcc_linux-64 8.9
+clang_osx-arm64 12
+msvc_win-64 19.29
+```
+
+The function `${{ compiler("foo") }}` thus evaluates to `{foo_compiler}_{target_platform} {foo_compiler_version}`.
+
+The variant config could look something like:
+
+```yaml
+c_compiler:
+  - gcc
+c_compiler_version:
+  - "8.9"
+cxx_compiler:
+  - clang
+cxx_compiler_version:
+  - "12"
+```
+
+> [!NOTE]
+> Default values for the `compiler` function
+>
+> If not further specified, the following values are used as default values:4
+>
+> ```yaml
+> linux:
+>  c: gcc
+>  cxx: gxx
+>  fortran: gfortran
+>  rust: rust
+> osx:
+>   c: clang
+>   cxx: clangxx
+>   fortran: gfortran
+>   rust: rust
+> win:
+>   c: vs2017
+>   cxx: vs2017
+>   fortran: gfortran
+>   rust: rust
+> ```
+
+## The `stdlib` function
+
+The `stdlib` function works exactly as the `compiler` function.
+
+## The `cdt` function
+
+CDT stands for "core dependency tree" packages. These are repackaged from Centos.
+
+The function expands to the following: 
+
+- package-name-<cdt_name>-<cdt_arch>
+
+Where `cdt_name` and `cdt_arch` are loaded from the variant config. If they are undefined, they default to:
+
+- `cos6` for `cdt_name` on `x86_64` and `x86`, otherwise `cos7`
+- To the `platform::arch` for `cdt_arch`, except for `x86` where it defaults to `i686`.
+
+## The pin functions
+
+The new recipe format has two `pin` expressions:
+
+- `pin_compatible`
+- `pin_subpackage`
+
+Both follow the same "pinning" mechanism as described next and have the same arguments.
+
+### Pin definition
+
+A pin has the following arguments:
+
+- `min_pin`: The lower bound of the dependency spec. This is expressed as a `x.x....` version where the `x` are filled in from the corresponding version of the package. For example, `x.x` would be `1.2` for a package version `1.2.3`. The resulting pin spec would look like `>=1.2` for the `min_pin` argument of `x.x` and a version of `1.2.3` for the package.
+- `max_pin`: This defines the upper bound and follows the same `x.x` semantics but adds `+1` to the last segment. For example, `x.x` would be `1.(2 + 1)` for a package version `1.2.3`. The resulting pin spec would look like `<1.3` for the `max_pin` argument of `x.x` and a version of `1.2.3` for the package.
+- `exact`: This is a boolean that specifies whether the pin should be exact. It defaults to `False`. If `exact` is `True`, the `min_pin` and `max_pin` are irrelevant. We pin to a `==` version and also include the build string exactly (e.g. `==1.2.3=h1234`).
+
+#### Example
+
+If we consider a package like `numpy-1.21.3-h123456_5` we could apply some pin expressions.
+
+- `min_pin=x.x, max_pin=x.x` would result in `>=1.21,<1.22`
+- `min_pin=x.x.x, max_pin=x` would result in `>=1.21.3,<2`
```

Comment:
Should the resolution of `max_pin` add the `.dev0` component too to avoid the dev versions leaking in in the version constraint?

Current behavior in solvers is that `<2` actually includes things like `v2.0b1`.

Reply:
Do you know what `conda-build` does?

I think the pre-release discussion is another one (maybe?). I would _really_ like to have some behavior like on PyPI where you can opt-in to pre-releases. But I think that's for another CEP.

Just for reference, I didn't try to invent something new with these pin expressions but rather just copy whatever `conda-build` does.

Reply:
I think that would be a good idea from the POV of least surprise. It happened in the numpy 2 migration that people were surprised that  `2.0.0rc2` actually satisfies `<2` (myself included).

Reply:
Care to write a CEP?! :) Would love to introduce that behavior.

Reply:
> Do you know what conda-build does?

Apparently it does handle this case. See [`gau2grid` recipe](https://github.com/conda-forge/gau2grid-feedstock/blob/32987c92c3933afe3da0dd21fec166599212d422/recipe/meta.yaml#L16) + its [rendered run_exports including the `a0` bit](https://conda-metadata-app.streamlit.app/?q=conda-forge%2Flinux-64%2Fgau2grid-2.0.7-h0b41bf4_2.conda).

The code in conda-build has it [here](https://github.com/conda/conda-build/blob/9c0ceec8d96d265e33d960c46024db78f4e3efd8/conda_build/utils.py#L1514). That should better be a `0dev0`, though, since that's technically the lowest bound (dev < alpha).